### ISSUE TITLE
Test on real hardware and measure resources consumptions

### ISSUE THREAD

**Issue by benoit74** [offspot/metrics#7]
Test on real hardware (Pi Zero at least) and confirm that system is OK.

Measure resources consumption (CPU, RAM) on this real hardware and confirm that it is acceptable or fix.

**Comment by benoit74** [comment]
@rgaudin
Is there a simple way to test this with the new hotspot stack based on Docker ? 
Or should I simply install the old stack and tweak it to make the service start and measure consumption ?

**Comment by rgaudin** [comment]
Install docker and your tools via docker to audit your tools' resources usage.
That should be enough to report metrics footprint. We'll have to test the whole thing altogether anyway, in time.

**Comment by benoit74** [comment]
Thank you

**Comment by benoit74** [comment]
Benchmarking tools are here: https://github.com/benoit74/benchmarking

**Comment by benoit74** [comment]
Discussed with rgaudin: I should test on PI3 (my old PI3 B are perfect, slower than PI3 A and B+) since we support only `armv8` / `arm64` (see https://github.com/offspot/base-image/issues/45 for details about this) and Pi Zero 2W has comparable performances.

**Comment by kelson42** [comment]
@rgaudin @benoit74 I would recommend to integrate this benchmarking tool to the future kiwix-hotspot benchmark repo.

**Comment by benoit74** [comment]
I agree that the tool used to monitor benchmarking operations (currently under https://github.com/benoit74/benchmarking) can be shared, it is just a light observability stack. This is why I asked where I should create a repo for this tool months ago. This should however be renamed, it is not doing any benchmark, only recording / analyzing node operation.

The benchmarking tool for offspot/metrics is in fact very different and specific to this system, since it is injecting traffic to the offspot/metrics system. This cannot be shared.

**Comment by benoit74** [comment]
Regarding benchmarking the offspot, I need to define a base dataset.

At this point, I think that it is especially important to define:
- the number of packages visited per year (package = 1 ZIM or 1 App or 1 set of static files)
- the number of package pages visited per year (page = more or less one web page, only for ZIMs, see [open question](https://github.com/offspot/metrics/issues/33))

I expect these two indicators are the main drivers of DB records numbers and other are only a second order effect (e.g. numbers of shared files created, real number of visits on each package or page, ...).

Do we have any idea of what would be a realistic "worst-case" scenario of these two values?

Nota: I chose the "per year" value to simplify a bit, HTH

**Comment by rgaudin** [comment]
About the number of packages per device, from the Cardshop, we have a median at 16, an average at 41 but 10% with over 100 packages, 3% over 200 and even 2 cards with over 1,000 packages.
Not sure how useful this is 🤷‍♂️

**Comment by benoit74** [comment]
It is already way better than only my non-educated guess ;)

**Comment by benoit74** [comment]
In order to generate data to stress test metrics, I have divided the procedure in two parts:
- a synthetic dataset of KPIs + Indicators 
- a log injector which injects raw Caddy logs

The synthetic dataset is useful to exercise the computation of KPIs on a huge dataset.

The log injector is useful to exercise the live processing capability of incoming events.

## Synthetic dataset

Datetime of the test system must be manipulated for perf tests so that we update KPIS on the 1st of January at midnight. Since we keep data for one year (from 1st of January to 31st of December), this will be the worst case scenario.

Indicators are generated to represent a whole year of data.

KPIs are generated for all aggregation periods except the last ones which will be anyway computed by the live system: 6 days, 3 weeks, 2 months. For years, I decided there would be 3 years already stored in DB.


### Indicators

Indicators are very important since they consist in the sheer size of the database + are mandatory for KPI refresh.

Indicators are kept for one year, so we need to generate one year worth of data.

Indicator values are numerous because we keep all data since we do not yet know what will be useful (e.g. we keep track of the number of visits on all pages we found, because we will need it to compute the top 50 but do not yet know which pages will be the top 50 - any page even with very low volume could become the top 1 on the last day of the year if the page receives a very huge traffic on this last day).

The volume of indicators is what will drive the load to refresh yearly, monthly, weekly and daily KPIs.

### KPIs

KPIs are not really useful since they are very small and almost do not impact performance, but this synthetic data will be very useful later on to exercise the UI with real data, so with still generate them.

### Data definition

We have defined an average yearly dataset which is used to generate the synthetic dataset. This average dataset is composed of sufficient data to generate indicators and KPIs. 

Indicator values are randomized to produce different but still coherent values around a baseline. 

KPI values are randomized as well, and scaled to adjust to the aggregation period (e.g. the monthly KPI will be 1/12th of the baseline, etc, ...).

Indicator dimensions (e.g. package name for popular packages) are partially defined in the average dataset. But for some data, we also need to generate some indicator dimensions randomly as well. E.g. for the PopularPages, we have some page names which have been defined in the average dataset and will be present in the top 50 (useful for UI testing for instance) but many pages with low visit volumes are generated and named for instance Page1, Page2, ... They are useful only to produce lots of data volume in the database. 

Main driver of the data volume is the number of ZIMs which drives the number of packages and the number of pages visited.

### Scenarios 

Multiple scenarios will be tested for the average yearly dataset to exercise the system. We can test scenarios with 16, 50, 100, 200, 500 and finally 1000 ZIMs. We won't go further than 1000 since it is probably not ran on a small Pi3 or most content is never accessed.

For every scenario, we will:
- create the synthetic dataset on a fast machine and transfer the DB to the test Pi 3
- start the benchmarking suite to record machine activity on the Pi3 (and wait few minutes so that we have a baseline of computing activity)
- position the clock just before midnight on 31 Dec
- start the metrics system (with processing activated)
- wait for the KPI update logic to trigger at midnight
- stop the metrics system
- transfer the benchmarking data to a fast machine and analyze the results
- observe the system performance at midnight where all KPIs (yearly, monthly, weekly and yearly will have been refreshed, with what is almost the biggest possible dataset)


## Log injection

Log injection consists in creating logs as Caddy would have produced them. They are then processed by metrics stack to create indicators (and then KPIs).

The log injection scenario is not meant to be ran on the log term (except to exercise the system stability). The main goal is to detect the system capability
to process logs on the fly. A 5 minutes test is hence way enough to observe how the system is capable to process the logs flow.

### Data definition

Log events are generated "manually", based on the Caddy log format and . We do not use real systems to simplify the testing procedure. We define the number of log events per kind of events (edupi file creation or deletion, edupi random request, ...). All these logs are generated in a temporary file and then randomized. This consumes a significant disk space but avoids to do all this in memory (which would probably never work on a Pi3). All logs are then inserted in the log file, which is also rotated as it would be in reality.

The load generated by the log injection is a function of:
- the number of log events generated for each kind of event
- an acceleration factor used to generate logs more kickly (with acceleration = 1, logs are equally spread on 24h, with acceleration = 4 they are spread on 6h)

The load is repeated forever (i.e. once we have reached the end of logs to generate, we start over).

### Scenarios 

Multiple scenarios will be tested, from 1 log every 5 secs to 1 log every 0.05 secs.

For every scenario, we will:
- start the benchmarking suite to record machine activity on the Pi3 (and wait few minutes so that we have a baseline of computing activity)
- start the metrics system (with processing activated)
- start the log injection
- stop the metrics system
- transfer the benchmarking data to a fast machine and analyze the results
- observe the system performance during log injection

**Comment by rgaudin** [comment]
Thank you for this detailed plan! Very insightful.

What's the expected output? What data will this produce and what are we gonna do with it? I see CPU and RAM usage. How will this be expressed?

In a real world scenario, this will run aside the actual apps and each process will compete for resources: RAM is finite (we have no swap) but CPU is dynamic. Usage might depend on external load. Just want to acknowledge this ; not something we can act upon.

My concern is I/O on disk (sdcard). In your metrics scope, you mostly read from disk (the log files) and that will have no impact but in the field, all apps will be reading/writing from/to the disk and this combination might slow things down. I imagine we only enable Caddy logs on a device if metrics has been requested.
Again, just food for thought.

**Comment by benoit74** [comment]
Synthetic dataset tests have been run on a Raspberry Pi 3B with commit https://github.com/offspot/metrics/commit/0cca4eb

We have created 10 datasets for these tests. Each contains 1 year of data, representing the worst case scenario in terms of DB size and performance test is launched on 31st Dec at 23:58 to launch end-of-year computation, again worst case scenario in terms of computation amount to perform in one go. No logs are generated during these tests.

During the tests, CPU has always been around 100% during the whole end-of-year computation. It was zero the rest of the time.

Details below. Quite good from my PoV (these are the worst case scenarii in terms of device performance and amount of data to process). WDYT?

| | TWENTY | FIFTY | HUNDRED_LIGHT | TWOHUNDRED_LIGHT | FIVEHUNDRED_LIGHT| THOUSAND_LIGHT | HUNDRED | TWOHUNDRED | FIVEHUNDRED | THOUSAND |  
|--|--|--|--|--|--|--|--|--|--|--|
| **Description** | Simulation of 20 ZIMs | Simulation of 50 ZIMs |  Simulation of 100 ZIMs, but mostly same traffic as FIFTY|  Simulation of 200 ZIMs, but mostly same traffic as FIFTY |  Simulation of 500 ZIMs, but mostly same traffic as FIFTY|  Simulation of 1000 ZIMs, but mostly same traffic as FIFTY | Simulation of 100 ZIMs |Simulation of 200 ZIMs |  Simulation of 500 ZIMs |Simulation of 1000 ZIMs |
| **DB Size** | 6.22 Mo | 10.37 Mo | 9.65 Mo |  10.88 Mo |  11.70 Mo | 13.18 Mo |  16.53 Mo | 28.98 Mo |  54.93 Mo |  107.07 Mo | 
| **IndicatorPeriod** | 3650 |3650 |3650 |3650 |3650 |3650 |3650 |3650 |3650 |3650 |
| **KpiRecord - Total** | 95 |95 |95 |95 |95 |95 |95 |95 |95 |95 |
| **KpiRecord - PackagePopularity**  | 19 | 19 | 19 | 19 | 19 | 19 | 19 | 19 | 19 | 19 |
| **KpiRecord - PopularPages**  | 19 | 19 | 19 | 19 | 19 | 19 | 19 | 19 | 19 | 19 |
| **KpiRecord - TotalUsage**  | 19 | 19 | 19 | 19 | 19 | 19 | 19 | 19 | 19 | 19 |
| **KpiRecord - Uptime**  | 19 | 19 | 19 | 19 | 19 | 19 | 19 | 19 | 19 | 19 |
| **KpiRecord - SharedFiles**  | 19 | 19 | 19 | 19 | 19 | 19 | 19 | 19 | 19 | 19 |
| **Indicator records - Total** |  99676 | 148972 | 151256 | 155650 | 168404 | 189574 | 232021 | 397327 | 730128 | 1396076 |
| **Indicator records - PackageHomeVisit** | 13583 | 14775 | 16883 | 21187 | 34026 | 54929 | 16883 | 21187 | 34026 | 54929 |
| **Indicator records - PackageItemVisit** | 79225 | 127326 | 127494  | 127575 | 127485 | 127773 | 208267 | 369273 | 689229 | 1334263 |
| **Indicator records - SharedFilesOperations** | 561 | 570 | 571  | 564 | 569 | 556 | 565 | 562 | 564 | 570 |
| **Indicator records - Uptime** |  3650 | 3650 | 3650  | 3650 | 3650 | 3650 | 3650 | 3650 | 3650 | 3650 |
| **Indicator records - TotalUsageOverall** |  880 | 870 | 875 | 881 | 876 | 882 | 886 | 866 | 874 | 870 |
| **Indicator records - TotalUsageByPackage** | 1777 | 1781 | 1783 | 1793 | 1798 | 1784 | 1770 | 1789 | 1785 | 1794 |
| **Indicator dimensions** | 4074 | 10104 | 10154 | 10254 | 10554 | 11054 | 20154 | 40254 | 80554 | 161054 |
| Max RSS used by container (Mo) | 135 | 128 | 143 | 127 | 127 | 129 | 161 | 136 | 166 | 169 |
| Runtime for end-of-year computation (s) | 4.5 | 6 | 6.5 | 8 | 7 | 8 | 13 | 17 | 46 | 117 |

**Comment by rgaudin** [comment]
What's the typical (idle/low traffic) RAM consumption of the deployed solution?

**Comment by rgaudin** [comment]
> I just realized that the Max RSS used is in fact used at startup (looks like it is alembic process who consumes this amount of RAM, that's weird).

OK

> 
> ![image](https://private-user-images.githubusercontent.com/7102089/287219857-f5a99105-e505-4465-94e7-3343c66574cc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE0MjI5NjcsIm5iZiI6MTcwMTQyMjY2NywicGF0aCI6Ii83MTAyMDg5LzI4NzIxOTg1Ny1mNWE5OTEwNS1lNTA1LTQ0NjUtOTRlNy0zMzQzYzY2NTc0Y2MucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDFUMDkyNDI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjA1MmFlNDU5NzBhNzk1OTI0ZjBjZWEwYmUwZmJkYzc2ZWYwOTllZmM2ZmNjNWJiZDljNjNlNjA1NDkzZmZlZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.FriMnIIXvLlcGCe8AUkSvvVeVf6idB8RzUJVghvvhX4)
> 

Nice ; please share how you produced those graphs as well (in the benchmark repo) as we'll need to test all offspot apps as well

> The idle RAM consumption after startup is around 85-86Mo and in fact we do not observe any significant change during the end of year computations, even for big ones like the "THOUSAND" case (picture above).

That's the most interesting IMO. We're not there yet but we ultimately want to pre-compute the ressource pressure of an offspot config to inform maker. For instance, this might be too much on a Pi0-2W _just for metrics_ (compared to content packages)

**Comment by benoit74** [comment]
> Nice ; please share how you produced those graphs as well (in the benchmark repo) as we'll need to test all offspot apps as well

This is just a snapshot of a small part of the "[Metricbeat Docker] Overview ECS" dashboard which is already featured in the benchmarking README.md (second picture + explicitly mentioned in the HOWTO section)

> That's the most interesting IMO. [...]

Yes. Once code has evolved and I decide to run synthetic datasets again, I will probably only run the "THOUSAND" dataset since it is sufficient to get a good grasp on `metrics` behavior for end-of-year computations and idle traffic. And of course I will still run the log injections scenario, but I do not yet know how interesting they are ^^

**Comment by benoit74** [comment]
I ran again the `THOUSAND` test case (worst one) with what will become the `v0.2.1` release, and it worked like a charm, the end-of-year computations took only 2 seconds since we know have **WAY** less data, since we have removed the "Popular Pages" KPI and associated indicator.

I also injected logs at high pace and there was no more concurrency issue, it looks like everything is progressing "smoothly". I struggled to generate a lot of delay in logs processing because when the system is at high load, even the log generator struggle to generate logs ... which is quite obvious. Nevertheless, I achieved to get into a situation where once I stopped the log generator, it took more or less 3 secs for metrics to recover, also proving that we are now really advancing the clock as log arrives. I don't see any interest to invest more time into finding a solution to generate even more logs, because anyway we will always be in a situation where we will have a competition for resources between metrics and the systems which generates the logs.

I finally let the system run for hours with a moderate volume of logs generated and nothing special happened.

I propose to close this issue, considering I have confirmed that for now the system is behaving properly on a small Pi device.
